03_Chatter/02-add-a-perisistence-layer.markdown

```diff
@@ -59,4 +62,86 @@ What this does is spawn an elixir process to do the save, outside of the request
 
 ### Display recent messages when a user joins the channel
 
-So, we have some messages saved, but it would be nice to see them when joining the channel. To do that, we have to change the `join/2` method to return a list of messages, and have the front end code parse and display those messages.
+So, we have some messages saved, but it would be nice to see them when joining the channel. To do that, we have to change the `join/3` method to return a list of messages, and have the front end code parse and display those messages.
+
+First, in order to complete the joining process, we have to send back `{:ok, socket}` from the `join/3` call. We can't send messages through the socket until the joining process has been completed. Instead, we'll send a call to the `gen_server` underlying our channel to send the messages _after_ the process has completed.
+
+We can handle this by using generic elixir with `send/2`. Let's update our `join/3` message to send the message of `:after_join` in our `join/3` method:
+
+```elixir
+def join("chat_room:lobby", payload, socket) do
+  if authorized?(payload) do
+    send(self, :after_join) # <~ add this line
+    {:ok, socket}
+  else
+    {:error, %{reason: "unauthorized"}}
+  end
+end
+```
+
+Now, we'll need to handle this message using a new handler called `handle_info/2`. Let's add this to our gen_server:
+
+```elixir
+def handle_info(:after_join, socket) do
+  {:noreply, socket}
+end
+```
+
+Now this `handle_info/2` will get called after the socket has been connected. We can do anything we want in here. Let's fetch the most recent messages and send them (one-by-one) to our client socket.
+
+### Fetching most recent messages
+
+Since we're going to be getting these messages a lot, it makes sense to put that code in the schema for messages, `lib/chatter/message.ex`. We can add the folloing method to our module, after the `changeset/2` method:
+
+```elixir
+def recent_messages(limit \\ 10) do
+  Chatter.Repo.all(Chatter.Message, limit: limit)
+end
+```
+
+This gives us a default of 10 messages, which will work well for our current case.
+
+Now, we can add this to our channel  `handle_info/3` call.
+
+```elixir
+def handle_info(:after_join, socket) do
+  Chatter.Message.recent_messages()
+  {:noreply, socket}
+end
+```
+
+Next, we need to broadcast the messages to the channel. `Enum.each/2` will let us step through each item in the collection and reply with the `push/3` method:
+
+```elixir
+def handle_info(:after_join, socket) do
+  Chatter.Message.recent_messages()
+  |> Enum.each(fn msg -> push(socket, "new_message", %{
+    name: msg.name,
+    message: msg.message
+  }) end)
+  {:noreply, socket}
+end
+```
+
+This should now work. When we reload the page, we should see some recent messages in the chat history. It's not a very neat method at this point though. Something easy to take out, since it's already in it's own closure, is the formatting of the message.
```

Comment:
This part seems unnecessary to me "Something easy to take out, since it's already in it's own closure, is the formatting of the message." maybe worth taking out?  It  seems to distract a bit from what you are trying to teach.